src/openmc_heat_driver.cpp

```diff
@@ -214,42 +368,73 @@ void OpenmcHeatDriver::set_heat_source()
   }
 }
 
+void OpenmcHeatDriver::update_density()
+{
+  if (this->has_global_coupling_data()) {
+    std::copy(densities_.begin(), densities_.end(), densities_prev_.begin());
+  }
+
+  densities_ = heat_driver_->density();
+  int solid_offset = heat_driver_->n_pins_ * heat_driver_->n_rings() * heat_driver_->n_axial_;
+
+  for (int i = 0; i < n_fluid_cells_; ++i) {
+    int index = i + n_solid_cells_;
+    const auto& c = openmc_driver_->cells_[index];
+    const auto& elems = cell_inst_to_elem_[index];
+
+    double average_rho = 0.0;
+    double total_vol = 0.0;
+    for (int elem_index : elems) {
+      double vol = heat_driver_->z_(elem_index + 1) - heat_driver_->z_(elem_index);
+      average_rho += densities_(elem_index + solid_offset);
+      total_vol += vol;
+    }
+
+    c.set_density(average_rho / total_vol);
+  }
+}
+
 void OpenmcHeatDriver::set_temperature()
 {
   const auto& r_fuel = heat_driver_->r_grid_fuel_;
   const auto& r_clad = heat_driver_->r_grid_clad_;
 
-  // For each OpenMC material, volume average temperatures and set
-  for (int i = 0; i < openmc_driver_->cells_.size(); ++i) {
-    // Get cell instance
+  // For each OpenMC solid cell, volume average solid temperatures and set
+  // by grabbing the T/H regions corresponding to the cell
+  for (int i = 0; i < n_solid_cells_; ++i) {
     const auto& c = openmc_driver_->cells_[i];
-
-    // Get rings corresponding to this cell instance
     const auto& rings = cell_inst_to_ring_[i];
 
-    // Get volume-average temperature for this material
     double average_temp = 0.0;
     double total_vol = 0.0;
     for (int ring_index : rings) {
-      // Use difference in r**2 as a proxy for volume. This is only used for
-      // averaging, so the absolute value doesn't matter
-      int j = ring_index % heat_driver_->n_rings();
-      double vol;
-      if (j < heat_driver_->n_fuel_rings_) {
-        vol = r_fuel(j + 1) * r_fuel(j + 1) - r_fuel(j) * r_fuel(j);
-      } else {
-        j -= heat_driver_->n_fuel_rings_;
-        vol = r_clad(j + 1) * r_clad(j + 1) - r_clad(j) * r_clad(j);
-      }
-
+      double vol = heat_driver_->solid_areas_(ring_index);
       average_temp += temperatures_(ring_index) * vol;
       total_vol += vol;
     }
-    average_temp /= total_vol;
 
-    // Set temperature for cell instance
+    average_temp /= total_vol;
     c.set_temperature(average_temp);
   }
+
+  // For each OpenMC fluid cell, set the fluid temperature by grabbing
+  // the T/H region corresponding to the cell
+  int solid_offset = heat_driver_->n_pins_ * heat_driver_->n_rings() * heat_driver_->n_axial_;
```

Comment:
This is a narrowing conversion, since the operands are size types.  Maybe declare `solid_offset` as `auto`?